django/contrib/admin/sites.py

```diff
@@ -245,20 +247,20 @@ def wrapper(*args, **kwargs):
 
         # Add in each model's views, and create a list of valid URLS for the
         # app_index
-        valid_app_labels = []
         for model, model_admin in self._registry.items():
+            base_path = r'%s/%s/' % (model._meta.app_label, model._meta.model_name)
             urlpatterns += [
-                url(r'^%s/%s/' % (model._meta.app_label, model._meta.model_name), include(model_admin.urls)),
+                url(r'^%s' % base_path, include(model_admin.urls)),
+                url(r'^%s(?P<field_name>.+)/$' % base_path,
+                    wrap(AutocompleteJsonView.as_view(admin_site=self)), name='autocomplete'),
             ]
-            if model._meta.app_label not in valid_app_labels:
-                valid_app_labels.append(model._meta.app_label)
 
         # If there were ModelAdmins registered, we should have a list of app
         # labels for which we need to allow access to the app_index view,
+        valid_app_labels = set(map(lambda x: x._meta.app_label, self._registry.keys()))
         if valid_app_labels:
-            regex = r'^(?P<app_label>' + '|'.join(valid_app_labels) + ')/$'
             urlpatterns += [
-                url(regex, wrap(self.app_index), name='app_list'),
+                url(r'^(?P<app_label>%s)/$' % '|'.join(valid_app_labels), wrap(self.app_index), name='app_list'),
```

Comment:
I see what you mean. I just copied the same behavior. Interestingly enough `normalise()` is from 2008 and the `|` part isn't even implemented. There still is a `fixme` comment.
Any idea on how to add a disjunct pattern? Or is there any reason not to support disjunct patterns?